File: src/superserve/cli/commands/deploy.py
"""CLI command for deploying agents."""

import io
import json
import shlex
import sys
import tarfile
import tempfile
import time
from pathlib import Path

import click
import yaml

from ..platform.client import PlatformAPIError, PlatformClient
from ..utils import Spinner, format_elapsed
from . import SUPERSERVE_YAML

# Directories and patterns excluded from the tarball
EXCLUDE_DIRS = {
    "__pycache__",
    ".git",
    ".venv",
    "venv",
    "node_modules",
    ".mypy_cache",
    ".pytest_cache",
    ".ruff_cache",
    "dist",
    "build",
}

EXCLUDE_FILE_PREFIXES = (".env",)


def _should_exclude(
    path: Path, root: Path, user_ignores: set[str] | None = None
) -> bool:
    """Check if a path should be excluded from the tarball."""
    rel = path.relative_to(root)
    parts = rel.parts

    # Check built-in directory exclusions
    for part in parts:
        if part in EXCLUDE_DIRS or part.endswith(".egg-info"):
            return True

    # Check built-in file exclusions
    if path.is_file() and any(path.name.startswith(p) for p in EXCLUDE_FILE_PREFIXES):
        return True

    # Check user-specified ignore patterns (matched as relative path prefixes)
    if user_ignores:
        rel_str = rel.as_posix()
        for pattern in user_ignores:
            if rel_str == pattern or rel_str.startswith(pattern + "/"):
                return True

    return False


def _make_tarball(project_dir: Path, user_ignores: set[str] | None = None) -> bytes:
    """Package a project directory into a tar.gz in memory."""
    buf = io.BytesIO()
    with tarfile.open(fileobj=buf, mode="w:gz") as tar:
        for item in sorted(project_dir.rglob("*")):
            if _should_exclude(item, project_dir, user_ignores):
                continue
            if item.is_file():
                arcname = str(item.relative_to(project_dir))
                tar.add(item, arcname=arcname)
    return buf.getvalue()


def _load_config(project_dir: Path) -> dict | None:
    """Load and validate superserve.yaml. Returns None if the file does not exist."""
    config_path = project_dir / SUPERSERVE_YAML
    if not config_path.exists():
        return None

    try:
        with open(config_path) as f:
            config = yaml.safe_load(f)
    except PermissionError:
        click.echo(f"Error: Permission denied reading {SUPERSERVE_YAML}.", err=True)
        sys.exit(1)
    except yaml.YAMLError as e:
        click.echo(f"Error: Invalid YAML in {SUPERSERVE_YAML}:\n  {e}", err=True)
        sys.exit(1)

    if not isinstance(config, dict):
        click.echo(f"Error: {SUPERSERVE_YAML} must be a YAML mapping.", err=True)
        sys.exit(1)

    if "name" not in config:
        click.echo(f"Error: 'name' is required in {SUPERSERVE_YAML}.", err=True)
        sys.exit(1)

    if "command" not in config:
        click.echo(f"Error: 'command' is required in {SUPERSERVE_YAML}.", err=True)
        sys.exit(1)

    return config


def _detect_command(entrypoint: str) -> str:
    """Auto-detect the run command from the entrypoint file extension."""
    ext = Path(entrypoint).suffix
    quoted = shlex.quote(entrypoint)
    commands = {
        ".py": f"python {quoted}",
        ".ts": f"npx tsx {quoted}",
        ".tsx": f"npx tsx {quoted}",
        ".js": f"node {quoted}",
        ".jsx": f"node {quoted}",
        ".mjs": f"node {quoted}",
        ".cjs": f"node {quoted}",
    }
    return commands.get(ext, f"python {quoted}")


def _format_size(size_bytes: int) -> str:
    """Format byte count as human-readable size."""
    if size_bytes >= 100 * 1024:
        return f"{size_bytes / (1024 * 1024):.1f} MB"
    return f"{size_bytes / 1024:.0f} KB"


def _write_temp_tarball(tarball_bytes: bytes) -> str:
    """Write tarball bytes to a temporary file and return the path."""
    with tempfile.NamedTemporaryFile(suffix=".tar.gz", delete=False) as tmp:
        tmp.write(tarball_bytes)
        return tmp.name


@click.command("deploy")
@click.argument("entrypoint", required=False, default=None)
@click.option(
    "--dir",
    "project_dir",
    default=".",
    type=click.Path(exists=True, file_okay=False, resolve_path=True),
    help="Project directory (default: current directory)",
)
@click.option(
    "--name",
    "-n",
    "agent_name",
    default=None,
    help="Agent name (defaults to directory name)",
)
@click.option("--port", "-p", default=None, type=int, help="Port for HTTP server mode")
@click.option("--json", "as_json", is_flag=True, help="Output as JSON")
def deploy(
    entrypoint: str | None,
    project_dir: str,
    agent_name: str | None,
    port: int | None,
    as_json: bool,
):
    """Deploy an agent to Superserve.

    Optionally pass an ENTRYPOINT file to deploy without superserve.yaml.

    \b
    Example:
        superserve deploy agent.py
        superserve deploy agent.py --name research-agent
        superserve deploy server.py --port 8000
        superserve deploy
        superserve deploy --dir ./my-agent
    """
    project_path = Path(project_dir)

    if entrypoint is not None:
        # Zero-config deploy: entrypoint provided as positional arg
        entrypoint_path = project_path / entrypoint
        if not entrypoint_path.exists():
            click.echo(
                f"Error: Entrypoint file '{entrypoint}' not found in {project_path}.",
                err=True,
            )
            sys.exit(1)

        name = agent_name or project_path.name
        command = _detect_command(entrypoint)
        mode = "http" if port else "shim"
        config: dict = {"entrypoint": entrypoint, "mode": mode}
        if port:
            config["port"] = port
        user_ignores: set[str] | None = None
    else:
        # Existing flow: load superserve.yaml
        loaded_config = _load_config(project_path)
        if loaded_config is None:
            click.echo(
                "Usage: superserve deploy <entrypoint> or create a superserve.yaml",
                err=True,
            )
            sys.exit(1)

        name = loaded_config["name"]
        command = loaded_config["command"]
        config = loaded_config
        user_ignores = set(loaded_config.get("ignore") or [])

    if as_json:
        tarball_bytes = _make_tarball(project_path, user_ignores)
        tarball_path = _write_temp_tarball(tarball_bytes)
        try:
            client = PlatformClient()
            agent = client.deploy_agent(
                name=name, command=command, config=config, tarball_path=tarball_path
            )
        except PlatformAPIError as e:
            click.echo(json.dumps({"error": e.message}), err=True)
            sys.exit(1)
        finally:
            Path(tarball_path).unlink(missing_ok=True)
        click.echo(json.dumps(agent.model_dump(), indent=2))
        return

    click.echo()
    deploy_start = time.time()

    # ── Package ──
    status = Spinner(indent=2)
    status.start("Packaging project...")
    tarball_bytes = _make_tarball(project_path, user_ignores)
    status.done(suffix=f"({_format_size(len(tarball_bytes))})")

    # ── Upload ──
    status.start("Uploading to Superserve...")
    tarball_path = _write_temp_tarball(tarball_bytes)

    try:
        client = PlatformClient()
        agent = client.deploy_agent(
            name=name,
            command=command,
            config=config,
            tarball_path=tarball_path,
        )
        status.done()
    except PlatformAPIError as e:
        status.fail()
        if e.status_code == 401:
            click.echo("Not authenticated. Run 'superserve login' first.", err=True)
        elif e.status_code == 422:
            click.echo(f"Validation error: {e.message}", err=True)
        else:
            click.echo(f"Error: {e.message}", err=True)
        sys.exit(1)
    finally:
        Path(tarball_path).unlink(missing_ok=True)

    # ── Dependencies ──
    if agent.deps_status == "installing":
        status.start("Installing dependencies...")
        poll_interval = 3
        max_wait = 300
        elapsed = 0
        deps_start = time.time()

        while elapsed < max_wait:
            time.sleep(poll_interval)
            elapsed += poll_interval

            try:
                agent = client.get_agent(name)
            except PlatformAPIError as e:
                status.fail()
                click.echo(f"Error checking status: {e.message}", err=True)
                click.echo(f"Check manually: superserve agents get {name}", err=True)
                sys.exit(1)

            if agent.deps_status == "ready":
                deps_time = format_elapsed(time.time() - deps_start)
                status.done(suffix=f"({deps_time})")
                break
            elif agent.deps_status == "failed":
                status.fail()
                if agent.deps_error:
                    click.echo(f"  {agent.deps_error}", err=True)
                click.echo(err=True)
                click.echo(
                    "Agent created but dependencies failed to install.", err=True
                )
                click.echo("Fix your requirements and run: superserve deploy", err=True)
                sys.exit(1)
        else:
            status.fail(suffix="(timed out)")
            click.echo(
                "\nDependency install is still running. Check status with:", err=True
            )
            click.echo(f"  superserve agents get {name}", err=True)
            sys.exit(1)

    # ── Done ──
    total_time = format_elapsed(time.time() - deploy_start)
    click.echo()
    click.echo(f"  Deployed '{agent.name}' in {total_time}")

    # ── Secrets ──
    required = config.get("secrets") or []
    if required:
        missing = [s for s in required if s not in agent.environment_keys]
        if missing:
            click.echo()
            click.echo("  Set your secrets before running:")
            for key in missing:
                click.echo(f"    superserve secrets set {name} {key}=...")
    elif not agent.environment_keys:
        click.echo()
        click.echo("  Set your API keys as secrets:")
        click.echo(f"    superserve secrets set {name} KEY=VALUE")

    click.echo()
    click.echo(f'  superserve run {agent.name} "your prompt here"')
    click.echo()

File: cli/src/commands/deploy.ts
import { randomUUID } from "node:crypto"
import { unlinkSync, writeFileSync } from "node:fs"
import { tmpdir } from "node:os"
import { join } from "node:path"
import { Command } from "commander"

import { track } from "../analytics"
import { createClient } from "../api/client"
import { PlatformAPIError } from "../api/errors"
import { loadProjectConfig } from "../config/project"
import { withErrorHandler } from "../errors"
import { commandBox } from "../utils/command-box"
import { formatElapsed, formatSize } from "../utils/format"
import { makeTarball } from "../utils/fs"
import { confirm } from "../utils/prompt"
import { createSpinner } from "../utils/spinner"

function writeTempTarball(data: Uint8Array): string {
  const path = join(tmpdir(), `superserve-${randomUUID()}.tar.gz`)
  writeFileSync(path, data)
  return path
}

export const deploy = new Command("deploy")
  .description("Deploy an agent to Superserve")
  .option("--dir <path>", "Project directory (default: current directory)", ".")
  .option("--json", "Output as JSON")
  .option("-y, --yes", "Skip confirmation")
  .action(
    withErrorHandler(
      async (options: { dir: string; json?: boolean; yes?: boolean }) => {
        const config = loadProjectConfig(options.dir)
        const client = createClient()
        const { name, command } = config

        // Check if agent already exists and confirm overwrite
        if (!options.json && !options.yes) {
          try {
            await client.getAgent(name)
            const ok = await confirm(
              `Agent '${name}' already exists. Redeploy?`,
            )
            if (!ok) {
              console.log("Cancelled")
              return
            }
          } catch (e) {
            if (!(e instanceof PlatformAPIError && e.statusCode === 404)) {
              throw e
            }
          }
        }

        const userIgnores = new Set(config.ignore ?? [])

        // JSON mode: simple output, no spinners
        if (options.json) {
          let tarballPath: string | undefined
          try {
            const tarballBytes = await makeTarball(options.dir, userIgnores)
            tarballPath = writeTempTarball(tarballBytes)
            const agent = await client.deployAgent(
              name,
              command,
              config,
              tarballPath,
            )
            await track("cli_deploy_success", {
              agent_name: name,
              tarball_bytes: tarballBytes.length,
            })
            console.log(JSON.stringify(agent, null, 2))
          } finally {
            if (tarballPath) {
              try {
                unlinkSync(tarballPath)
              } catch {}
            }
          }
          return
        }

        // Interactive mode with spinners
        console.log()
        const deployStart = performance.now()
        const status = createSpinner({ indent: 2 })

        // Package
        status.start("Packaging project...")
        const tarballBytes = await makeTarball(options.dir, userIgnores)
        status.done("\u2713", `(${formatSize(tarballBytes.length)})`)

        // Upload
        status.start("Uploading to Superserve...")
        const tarballPath = writeTempTarball(tarballBytes)

        let agent
        try {
          agent = await client.deployAgent(name, command, config, tarballPath)
          status.done()
        } finally {
          try {
            unlinkSync(tarballPath)
          } catch {}
        }

        // Wait for dependency installation (sandbox template build)
        if (agent.sandbox_status === "building") {
          status.start("Installing dependencies...")
          const pollInterval = 3000
          const maxWait = 300_000
          let elapsed = 0
          const buildStart = performance.now()

          while (elapsed < maxWait) {
            await Bun.sleep(pollInterval)
            elapsed += pollInterval

            agent = await client.getAgent(name)

            if (agent.sandbox_status === "ready") {
              status.done(
                "\u2713",
                `(${formatElapsed((performance.now() - buildStart) / 1000)})`,
              )
              break
            }
            if (agent.sandbox_status === "failed") {
              status.fail()
              await track("cli_deploy_build_failed", { agent_name: name })
              console.error()
              console.error("Agent created but dependency install failed.")
              console.error("Fix your requirements and run:")
              console.error(commandBox("superserve deploy"))
              process.exitCode = 1
              return
            }
          }

          if (agent.sandbox_status === "building") {
            status.fail("(timed out)")
            await track("cli_deploy_build_timeout", { agent_name: name })
            console.error(
              "\nDependency install is still running. Check status with:",
            )
            console.error(commandBox(`superserve agents get ${name}`))
            process.exitCode = 1
            return
          }
        }

        // Done
        const totalSeconds = (performance.now() - deployStart) / 1000
        const totalTime = formatElapsed(totalSeconds)
        await track("cli_deploy_success", {
          agent_name: name,
          tarball_bytes: tarballBytes.length,
          duration_s: Math.round(totalSeconds),
        })
        console.log()
        console.log(`  Deployed '${agent.name}' in ${totalTime}`)

        // Secrets
        const required = config.secrets ?? []
        if (required.length > 0) {
          const missing = required.filter(
            (s) => !agent.environment_keys.includes(s),
          )
          if (missing.length > 0) {
            console.log()
            console.log("  Set your secrets before running:")
            for (const key of missing) {
              console.log(
                commandBox(`superserve secrets set ${config.name} ${key}=...`),
              )
            }
          }
        } else if (agent.environment_keys.length === 0) {
          console.log()
          console.log("  Set your API keys as secrets:")
          console.log(
            commandBox(`superserve secrets set ${config.name} KEY=VALUE`),
          )
        }

        console.log()
        console.log(
          commandBox(`superserve run ${agent.name} "your prompt here"`),
        )
        console.log()
      },
    ),
  )
